internal/fs/dir_handle_test.go

```diff
@@ -0,0 +1,272 @@
+// foo/bar   --File
+// fetchEntriesAsync will return 1 entry for directory with 1 file.
+func (t *DirHandleTest) FetchAsyncEntries_NonEmptyDir() {
+	t.createSimpleDirectory()
+	t.createDirHandle(false, false, dirInodeName)
+	t.dh.FetchEntriesAsync(fuseops.RootInodeID)
+	entries := t.dh.entries
+
+	AssertEq(1, len(entries))
+	AssertEq(fileUnderDir, entries[0].Name)
+	AssertEq(true, t.dh.entriesValid)
+	t.resetDirHandle()
+}
+
+// Directory Structure Used
+// foo              --Directory
+// foo/baz          --Implicit Directory
+// foo/baz/bar      --file
+// fetchEntriesAsync will return 1 entry for implicit directory if flag is set to true.
+func (t *DirHandleTest) FetchAsyncEntries_ImplicitDir_FlagTrue() {
+	err := t.createImplicitDirDefinedByFile()
+	AssertEq(nil, err)
+
+	//implicit-dirs flag set to true
+	t.createDirHandle(true, false, dirInodeName)
+	t.dh.FetchEntriesAsync(fuseops.RootInodeID)
+	entries := t.dh.entries
+
+	AssertEq(1, len(entries))
+	AssertEq(implicitDirName, entries[0].Name)
+	AssertEq(true, t.dh.entriesValid)
+	t.resetDirHandle()
+}
+
+// Same directory structure as above.
+// fetchEntriesAsync will return 0 entry for implicit directory if flag is set to false.
+func (t *DirHandleTest) FetchAsyncEntries_ImplicitDir_FlagFalse() {
+	err := t.createImplicitDirDefinedByFile()
+	AssertEq(nil, err)
+
+	//implicit-dirs flag set to false
+	t.createDirHandle(false, false, dirInodeName)
+	t.dh.FetchEntriesAsync(fuseops.RootInodeID)
+
+	AssertEq(0, len(t.dh.entries))
+	AssertEq(true, t.dh.entriesValid)
+	t.resetDirHandle()
+}
+
+// Directory Structure Used
+// foo       --Directory
+// foo/bar   --File
+// Offset for first read dir operation is zero.Hence, entries for dir will
+// be fetched. Bytes read for this operation will be 32(size of 1 dirent)
+// as according to the directory structure created.
+func (t *DirHandleTest) Readdir_OffsetZero() {
+	t.createSimpleDirectory()
+	t.createDirHandle(false, false, dirInodeName)
+	op := &fuseops.ReadDirOp{
+		Inode:  fuseops.InodeID(dummyInodeId),
+		Handle: fuseops.HandleID(dummyHandleId),
+		Offset: fuseops.DirOffset(0),
+		OpContext: fuseops.OpContext{
+			FuseID: dummyfuseid,
+			Pid:    dummypid,
+		},
+		Dst: make([]byte, 1024),
+	}
+
+	err := t.dh.ReadDir(t.ctx, op)
+
+	AssertEq(nil, err)
+	AssertEq(1024, len(op.Dst))
+	AssertEq(32, op.BytesRead)
+}
+
+// Directory Structure Used
+// foo       --Directory
+// foo/bar   --File
+// Offset for first read dir operation is non-zero.Hence, no entries are fetched.
+// Bytes read for this operation will be zero.
+func (t *DirHandleTest) Readdir_OffsetNonZero() {
+	t.createSimpleDirectory()
+	t.createDirHandle(false, false, dirInodeName)
+	op := &fuseops.ReadDirOp{
+		Inode:  fuseops.InodeID(dummyInodeId),
+		Handle: fuseops.HandleID(dummyHandleId),
+		Offset: fuseops.DirOffset(1),
+		OpContext: fuseops.OpContext{
+			FuseID: dummyfuseid,
+			Pid:    dummypid,
+		},
+		Dst: make([]byte, 1024),
+	}
+
+	err := t.dh.ReadDir(t.ctx, op)
```

Comment:
can you also add a test when REadDir is triggered twice asking for 1 entry at a time

Ideally i want a test case for validating if 2nd page call is working fine or not. but given that we are fetching 5000 entries, its not feasible to write.

Reply:
Added test and its description.